### ISSUE TITLE
REST API layer to support headless use of Raytha

### ISSUE THREAD

**Issue by apexdodge** [RaythaHQ/raytha#9]
This is a non-trivial task, but is essential to build it well for v1.0.0.

For now we have a single API call available to allow developers to ping their application for health check purposes.

![image](https://user-images.githubusercontent.com/777005/210150045-ea5e9193-0cc8-4cfc-a952-5b6f330f8083.png)

The initial implementation of the REST API can be a CRUD functionality on:

- Users [read,write]
- UserGroups [read,write]
- Content Items [read,write]
- Media Items [read,write]
- Content Types [read only]

These will go a long way to get started and provide immediate value to devs. In later versions we can consider adding CRUD around Templates and other enhancements.

Some big considerations:
- Authentication method? Most preferred is via API key attached to an admin account. An admin account is assigned to roles with permissions that the API key will inherit.
- Code needs to be compatible with NSwag to autogenerate swagger API specs and UI

**Comment by apexdodge** [comment]
I think the most appropriate implementation is to add a new section above 'Delete admin' called API Keys

![image](https://user-images.githubusercontent.com/777005/213819861-b1c27c05-4e7e-4358-826a-ef717653af6b.png)

The admin can add or remove api keys that will be GUIDs.

The GUIDs will be shown to the admin only once and non-recoverable because they will be stored as salted and hashed values, like passwords.

Authentication to the api will use custom header:

`X-API-KEY`

If there is a match on the X-API-KEY, check the admin's roles and permissions and verify that they have permission to read or write to that resource.

- if API Key not found, throw a 401 
- if API key is found but user does not have the rights, throw a 403

**Comment by apexdodge** [comment]
Foundation built out. Remaining tasks:

- Implement some automapping since API is returning too much data.
- Remaining api calls
- `RaythaApiContentTypeAuthorizationHandler` and `RaythaAdminContentTypeAuthorizationHandler` may not be needed

**Comment by apexdodge** [comment]
Initial implementation built.

![FireShot Capture 064 - Swagger UI - localhost](https://user-images.githubusercontent.com/777005/215289240-4be1f07a-b3a4-458b-b2a1-388f4479f71d.png)

QA items:

- [ ] Better exception handling up and down the request pipeline
- [ ] Refactor the JsonConverters and SchemaFilters
- [ ] Use a proper model for returning on BadRequest
- [ ] MediaItems permission should allow MediaItem creation if the role has *any* content type with edit permission. Doesn't make sense otherwise.
- [ ] Did attachment field type have Choices in its definition unnecessarily?

**Comment by apexdodge** [comment]
Merged into `dev` branch!
